pithos/GstPlayer.py

```diff
@@ -0,0 +1,530 @@
+        nick='actual-state prop',
+        blurb='actual state of GstPlayer',
+        flags=GObject.ParamFlags.READABLE,
+    )
+    def actual_state(self):
+        return self._actual_state
+
+    @GObject.Property(
+        type=GObject.TYPE_PYOBJECT,
+        nick='desired-state prop',
+        blurb='desired state of GstPlayer',
+        flags=GObject.ParamFlags.READABLE,
+    )
+    def desired_state(self):
+        return self._desired_state
+
+    def plugins_installation_in_progress(self):
+        return GstPbutils.install_plugins_installation_in_progress()
+
+    def query_position(self):
+      if self.query(self.__query_position):
+          return self.__query_position.parse_position()[1]
+      else:
+          return 0
+
+    def start_stream(self, url):
+        self.props.location = url
+        self._set_player_state(PseudoGst.BUFFERING)
+
+    def play(self):
+        self._set_player_state(PseudoGst.PLAYING)
+
+    def pause(self):
+        self._set_player_state(PseudoGst.PAUSED)
+
+    def stop(self):
+        self._set_player_state(PseudoGst.STOPPED, change_gst_state=True)
+
+    def _query_duration(self):
+      if self.query(self.__query_duration):
+          return self.__query_duration.parse_duration()[1]
+
+    def _query_buffer(self):
+        if self.query(self.__query_buffer):
+            return self.__query_buffer.parse_buffering_percent()[0]
+        else:
+            return True
+
+    def _on_gst_stream_start(self, bus, message):
+        self.emit('new-stream-duration', self._query_duration())
+
+    def _on_gst_error(self, bus, message):
+        err, debug = message.parse_error()
+        self.emit('warning', (err, debug, err.code))
+
+    def _on_gst_plugin_installed(self, result, userdata):
+        if result == GstPbutils.InstallPluginsReturn.SUCCESS:
+            msg = (
+                _('Codec installation successful'),
+                _('The required codec was installed, please restart Pithos.'),
+            )
+            self.emit('fatal-error', msg)
+        else:
+            msg = (
+                _('Codec installation failed'),
+                None,
+                _('The required codec failed to install.\nEither manually install it or try another quality setting.'),
+            )
+            self.emit('error', msg)
+
+    def _on_gst_element(self, bus, message):
+        if GstPbutils.is_missing_plugin_message(message):
+            if GstPbutils.install_plugins_supported():
+                details = GstPbutils.missing_plugin_message_get_installer_detail(message)
+                GstPbutils.install_plugins_async([details,], None, self.on_gst_plugin_installed, None)
+            else:
+                msg = (
+                    _('Missing codec'),
+                    None,
+                    _('GStreamer is missing a plugin and it could not be automatically installed.'
+                            '\nEither manually install it or try another quality setting.'),
```

Comment:
__[flake8]__
* [E127] continuation line over-indented for visual indent

<!-- Comment by [SideCI](https://sideci.com) -->